File: kotlin/src/main/kotlin/advertisement/domain/model/value_object/AdvertisementDate.kt
package advertisement.domain.model.value_object

import arrow.core.Either
import arrow.core.right
import java.time.LocalDateTime

class AdvertisementDate private constructor(private val value: LocalDateTime) {

    companion object {
        fun build(value: LocalDateTime): Either<Unit, AdvertisementDate> {
            return AdvertisementDate(value).right()
        }
    }

    fun value(): LocalDateTime {
        return value
    }
}

File: typescript/src/advertisement/domain/model/value-object/AdvertisementDate.ts
import * as E from '@effect-ts/core/Either';

export class AdvertisementDate {
    private constructor(
      readonly _value: Date,
    ) {}

    static build(value: Date): E.Either<never, AdvertisementDate> {
        return E.right(new AdvertisementDate(value));
    }

    public value(): Date {
        return this._value;
    }
}